libs/utils/include/utils/RangeMap.h

```diff
@@ -0,0 +1,309 @@
+        add(begin, ++key, value);
+    }
+
+    /**
+     * Checks if a range exists that encompasses the given key.
+     */
+    bool has(KeyType key) const noexcept {
+        return find(key) != mMap.end();
+    }
+
+    /**
+     * Retrieves the element at the given location, panics if no element exists.
+     */
+    const ValueType& get(KeyType key) const {
+        const_iterator iter = find(key);
+        ASSERT_PRECONDITION(iter != end(), "RangeMap: No element exists at the given key.");
+        return get_value(iter);
+    }
+
+    /**
+     * Removes all elements between begin (inclusive) and end (exclusive).
+     */
+    void clear(KeyType first, KeyType last) noexcept {
+        // Check if an existing range contains "first".
+        iterator iter = find(first);
+        if (iter != end()) {
+            Range<KeyType>& existing = get_range(iter);
+            // Split the existing range into two ranges.
+            if (last < existing.last && first > existing.first) {
+                const KeyType tmp = existing.last;
+                existing.last = first;
+                insert(last, tmp, get_value(iter));
+                return;
+            }
+            // Clip the end of the existing range and potentially remove it.
+            existing.last = first;
+            clean(iter);
+            wipe(first, last);
+            return;
+        }
+
+        // Check if an existing range contains the end of the new range.
+        KeyType back = last;
+        iter = find(--back);
+        if (iter == end()) {
+            wipe(first, last);
+            return;
+        }
+        Range<KeyType>& existing = get_range(iter);
+
+        // Clip the beginning of the existing range and potentially remove it.
+        existing.first = last;
+        clean(iter);
+        wipe(first, last);
+    }
+
+    /**
+     * Shorthand for the "clear" method that clears a single element.
+     */
+    void reset(KeyType key) noexcept {
+        KeyType begin = key;
+        clear(begin, ++key);
+    }
+
+    /**
+     * Returns the number of internal interval objects (rarely used).
+     */
+    size_t range_count() const noexcept { return mMap.size(); }
+
+private:
+
+    using Map = std::map<KeyType, std::pair<Range<KeyType>, ValueType>>;
+    using iterator = typename Map::iterator;
+    using const_iterator = typename Map::const_iterator;
+
+    const_iterator begin() const noexcept { return mMap.begin(); }
+    const_iterator end() const noexcept { return mMap.end(); }
+
+    iterator begin() noexcept { return mMap.begin(); }
+    iterator end() noexcept { return mMap.end(); }
+
+    Range<KeyType>& get_range(iterator iter) const { return iter->second.first; }
+    ValueType& get_value(iterator iter) const { return iter->second.second; }
+
+    const Range<KeyType>& get_range(const_iterator iter) const { return iter->second.first; }
+    const ValueType& get_value(const_iterator iter) const { return iter->second.second; }
```

Comment:
this naming convention is not filament's. This should be `getRange()`, etc...

Reply:
Fixed.